components/hab/src/command/pkg/download.rs

```diff
@@ -0,0 +1,439 @@
+                target: PackageTarget,
+                download_path: Option<&PathBuf>,
+                token: Option<&str>,
+                verify: bool)
+                -> Result<()>
+    where U: UIWriter
+{
+    debug!("Starting download with url: {}, channel: {}, product: {}, version: {}, target: {}, \
+            download_path: {:?}, token: {:?}, verify: {}",
+           url, channel, product, version, target, download_path, token, verify);
+
+    let download_path_default = &cache_root_path::<PathBuf>(None); // Satisfy E0716
+    let download_path_expanded = download_path.unwrap_or(download_path_default).as_ref();
+    debug!("Using download_path {:?} expanded to {:?}",
+           download_path, download_path_expanded);
+
+    // We deliberately use None to specify the default path as this is used for cert paths, which
+    // we don't want to override.
+    let api_client = Client::new(url, product, version, None)?;
+    let task = DownloadTask { idents,
+                              target,
+                              url,
+                              api_client,
+                              token,
+                              channel,
+                              download_path: download_path_expanded,
+                              verify };
+
+    let download_count = task.execute(ui)?;
+
+    debug!("Expanded package count: {}", download_count);
+
+    Ok(())
+}
+
+struct DownloadTask<'a> {
+    idents:        Vec<PackageIdent>,
+    target:        PackageTarget,
+    url:           &'a str,
+    api_client:    BoxedClient,
+    token:         Option<&'a str>,
+    channel:       &'a ChannelIdent,
+    download_path: &'a Path,
+    verify:        bool,
+}
+
+impl<'a> DownloadTask<'a> {
+    fn execute<T>(&self, ui: &mut T) -> Result<usize>
+        where T: UIWriter
+    {
+        // This was written intentionally with an eye towards data parallelism
+        // Any or all of these phases should naturally fit a fork-join model
+
+        ui.begin(format!("Preparing to resolve packages and download transitive \
+                          dependendencies for {} package idents",
+                         self.idents.len()))?;
+        ui.begin(format!("Using channel {} from {}", self.channel, self.url))?;
+        ui.begin(format!("Storing in download directory {:?} ", self.download_path))?;
+
+        self.verify_and_prepare_download_directory(ui)?;
+
+        // Phase 1: Expand to fully qualified deps and TDEPS
+        let expanded_idents = self.expand_sources(ui)?;
+
+        // Phase 2: Download artifacts
+        let downloaded_artifacts = self.download_artifacts(ui, &expanded_idents)?;
+
+        Ok(downloaded_artifacts.len())
+    }
+
+    // For each source, use the builder/depot to expand it to a fully qualifed form
+    // The same call gives us the TDEPS, add those as
+    fn expand_sources<T>(&self, ui: &mut T) -> Result<HashSet<(Box<PackageIdent>, PackageTarget)>>
```

Comment:
I'm not sure why Box is needed here... should be able to eliminate... at least worth a TODO...

Reply:
:+1: